File: day_03/solve.js
const fs = require("fs");
const readline = require("readline");

const compute = ({ right = 3, down = 1, input = "./input" } = {}) => {
  const readInterface = readline.createInterface({
    input: fs.createReadStream(input),
    console: false,
  });

  return new Promise((resolve) => {
    let col = 0,
      lineNo = 0,
      trees = -1;
    readInterface.on("line", (line) => {
      if (trees == -1) return (trees = 0);
      if (++lineNo % down != 0) return;
      trees += line[(col += right) % line.length] == "#" ? 1 : 0;
    });
    readInterface.on("close", () => resolve(trees));
  });
};

const firstTask = async () => console.log(`First part: ${await compute()}`);
const secondTask = async () => {
  let result = await Promise.all(
    [
      { right: 1, down: 1 },
      { right: 3, down: 1 },
      { right: 5, down: 1 },
      { right: 7, down: 1 },
      { right: 1, down: 2 },
    ].map(({ right, down }) => compute({ right, down, input: "./input" }))
  ).then((r) => r.reduce((p, c) => p * c, 1));

  console.log(`Second part: ${result}`);
};

firstTask() && secondTask();

File: day_03/solve.exs
tree? = fn str, i -> String.at(str, rem(i, String.length(str))) == "#" end

solve = fn {rd, cd} ->
  {trees, _} =
    File.stream!("./input")
    |> Stream.map(&String.trim/1)
    |> Stream.with_index()
    |> Enum.reduce({0, 0}, fn
      _, {0, 0} ->
        {0, cd}

      {l, row}, {t, col} when rem(row, rd) == 0 ->
        {t + if(tree?.(l, col), do: 1, else: 0), col + cd}

      _, {t, col} ->
        {t, col}
    end)

  trees
end

IO.puts("First part: #{solve.({1, 3})}")

IO.puts(
  "Second part: " <>
    ([
       {1, 1},
       {1, 3},
       {1, 5},
       {1, 7},
       {2, 1}
     ]
     |> Enum.map(&solve.(&1))
     |> Enum.reduce(1, &(&1 * &2))
     |> to_string())
)
